<h5>URL Assignment: https://colab.research.google.com/drive/1T9HAtxKs9LS7xXHb0OmFNWbDOf1an6RG?usp=sharing#scrollTo=VuVSlxaZDqsj</h5>

In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv("IMDB Dataset.csv")

In [3]:
df

,review,sentiment
0,One of the other reviewers has mentioned that ...,positive
1,A wonderful little production. <br /><br />The...,positive
2,I thought this was a wonderful way to spend ti...,positive
3,Basically there's a family where a little boy ...,negative
4,"Petter Mattei's ""Love in the Time of Money"" is...",positive
...,...,...
49995,I thought this movie did a down right good job...,positive
49996,"Bad plot, bad dialogue, bad acting, idiotic di...",negative
49997,I am a Catholic taught in parochial elementary...,negative
49998,I'm going to have to disagree with the previou...,negative


In [4]:
df['review'][0]

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.<br /><br />The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word.<br /><br />It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many..Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.<br /><br />I would say the main appeal of the show is due to the fa

<h1>Text Preprocessing</h1>

<h5>Lowercasing</h5>

In [5]:
df['review']=df['review'].str.lower()

In [6]:
import re
def remove_html(review):
    pattern=re.compile('<.*?>')
    return pattern.sub(r'',review)

<h5>Removing HTML tags</h5>

In [7]:
df['review']=df['review'].apply(remove_html)

In [8]:
df['review'][0]

"one of the other reviewers has mentioned that after watching just 1 oz episode you'll be hooked. they are right, as this is exactly what happened with me.the first thing that struck me about oz was its brutality and unflinching scenes of violence, which set in right from the word go. trust me, this is not a show for the faint hearted or timid. this show pulls no punches with regards to drugs, sex or violence. its is hardcore, in the classic use of the word.it is called oz as that is the nickname given to the oswald maximum security state penitentary. it focuses mainly on emerald city, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. em city is home to many..aryans, muslims, gangstas, latinos, christians, italians, irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.i would say the main appeal of the show is due to the fact that it goes where other shows wo

<h5>Handling Contractions</h5>

In [9]:
import sys
!{sys.executable} -m pip install contractions

In [10]:
import contractions
def expand_contractions(text):
    return contractions.fix(text)

In [11]:
text = "I don't think this movie is good. Who'll watch it?"
print(expand_contractions(text))

I do not think this movie is good. Who will watch it?


In [12]:
df['review']=df['review'].apply(expand_contractions)

In [13]:
df['review'][0]

'one of the other reviewers has mentioned that after watching just 1 oz episode you will be hooked. they are right, as this is exactly what happened with me.the first thing that struck me about oz was its brutality and unflinching scenes of violence, which set in right from the word go. trust me, this is not a show for the faint hearted or timid. this show pulls no punches with regards to drugs, sex or violence. its is hardcore, in the classic use of the word.it is called oz as that is the nickname given to the oswald maximum security state penitentary. it focuses mainly on emerald city, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. them city is home to many..aryans, muslims, gangstas, latinos, christians, italians, irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.i would say the main appeal of the show is due to the fact that it goes where other show

<h5>Removing Punctuations</h5>

In [14]:
import string
exclude=string.punctuation

def remove_punc(review):
    return review.translate(str.maketrans(exclude," " * len(exclude)))

#removing puncs and adding spaces in place of them.

In [15]:
df['review']=df['review'].apply(remove_punc)

In [16]:
df['review'][0]

'one of the other reviewers has mentioned that after watching just 1 oz episode you will be hooked  they are right  as this is exactly what happened with me the first thing that struck me about oz was its brutality and unflinching scenes of violence  which set in right from the word go  trust me  this is not a show for the faint hearted or timid  this show pulls no punches with regards to drugs  sex or violence  its is hardcore  in the classic use of the word it is called oz as that is the nickname given to the oswald maximum security state penitentary  it focuses mainly on emerald city  an experimental section of the prison where all the cells have glass fronts and face inwards  so privacy is not high on the agenda  them city is home to many  aryans  muslims  gangstas  latinos  christians  italians  irish and more    so scuffles  death stares  dodgy dealings and shady agreements are never far away i would say the main appeal of the show is due to the fact that it goes where other show

<h5>Removing Extra Spaces</h5>

In [17]:
def remove_extra_spaces(text):
    return " ".join(text.split())

In [18]:
df['review']=df['review'].apply(remove_extra_spaces)

In [19]:
df['review'][0]

'one of the other reviewers has mentioned that after watching just 1 oz episode you will be hooked they are right as this is exactly what happened with me the first thing that struck me about oz was its brutality and unflinching scenes of violence which set in right from the word go trust me this is not a show for the faint hearted or timid this show pulls no punches with regards to drugs sex or violence its is hardcore in the classic use of the word it is called oz as that is the nickname given to the oswald maximum security state penitentary it focuses mainly on emerald city an experimental section of the prison where all the cells have glass fronts and face inwards so privacy is not high on the agenda them city is home to many aryans muslims gangstas latinos christians italians irish and more so scuffles death stares dodgy dealings and shady agreements are never far away i would say the main appeal of the show is due to the fact that it goes where other shows would not dare forget p

<h5>Stopword Removal</h5>

In [20]:
import nltk
from nltk.corpus import stopwords

In [21]:
def remove_stopwords(text):
    new_text=[]
    for word in text.split():
        if word in stopwords.words("english"):
            new_text.append(" ")
        else:
            new_text.append(word)
    x=new_text[:]
    new_text.clear()
    return " ".join(x)

In [22]:
df['review']=df['review'].apply(remove_stopwords)

In [23]:
df['review'][0]

'one       reviewers   mentioned     watching   1 oz episode       hooked     right       exactly   happened       first thing   struck     oz     brutality   unflinching scenes   violence   set   right     word go trust           show     faint hearted   timid   show pulls   punches   regards   drugs sex   violence     hardcore     classic use     word     called oz         nickname given     oswald maximum security state penitentary   focuses mainly   emerald city   experimental section     prison       cells   glass fronts   face inwards   privacy     high     agenda   city   home   many aryans muslims gangstas latinos christians italians irish       scuffles death stares dodgy dealings   shady agreements   never far away   would say   main appeal     show   due     fact     goes     shows would   dare forget pretty pictures painted   mainstream audiences forget charm forget romance oz     mess around   first episode   ever saw struck       nasty     surreal   could   say     ready 

In [24]:
df['review']=df['review'].apply(remove_extra_spaces)

<h5>Tokenization</h5>

In [27]:
from nltk.tokenize import word_tokenize

In [28]:
def tokenize(text):
    return word_tokenize(text)

In [29]:
df['review']=df['review'].apply(tokenize)

In [30]:
df['review'][0]

['one',
 'reviewers',
 'mentioned',
 'watching',
 '1',
 'oz',
 'episode',
 'hooked',
 'right',
 'exactly',
 'happened',
 'first',
 'thing',
 'struck',
 'oz',
 'brutality',
 'unflinching',
 'scenes',
 'violence',
 'set',
 'right',
 'word',
 'go',
 'trust',
 'show',
 'faint',
 'hearted',
 'timid',
 'show',
 'pulls',
 'punches',
 'regards',
 'drugs',
 'sex',
 'violence',
 'hardcore',
 'classic',
 'use',
 'word',
 'called',
 'oz',
 'nickname',
 'given',
 'oswald',
 'maximum',
 'security',
 'state',
 'penitentary',
 'focuses',
 'mainly',
 'emerald',
 'city',
 'experimental',
 'section',
 'prison',
 'cells',
 'glass',
 'fronts',
 'face',
 'inwards',
 'privacy',
 'high',
 'agenda',
 'city',
 'home',
 'many',
 'aryans',
 'muslims',
 'gangstas',
 'latinos',
 'christians',
 'italians',
 'irish',
 'scuffles',
 'death',
 'stares',
 'dodgy',
 'dealings',
 'shady',
 'agreements',
 'never',
 'far',
 'away',
 'would',
 'say',
 'main',
 'appeal',
 'show',
 'due',
 'fact',
 'goes',
 'shows',
 'would',
 

<h5>Lemmatization</h5>

In [20]:
from nltk.stem import WordNetLemmatizer
lemmatizer=WordNetLemmatizer()

In [21]:
def lemmatize_text(text):
    return " ".join(
        lemmatizer.lemmatize(word,pos="v")
        for word in text
    )

In [ ]:
df['review']=df['review'].apply(lemmatize_text)

In [3]:
df['review'][0]

'one reviewers mention watch 1 oz episode hook right exactly happen first thing strike oz brutality unflinching scenes violence set right word go trust show faint hearted timid show pull punch regard drug sex violence hardcore classic use word call oz nickname give oswald maximum security state penitentary focus mainly emerald city experimental section prison cells glass front face inwards privacy high agenda city home many aryans muslims gangstas latinos christians italians irish scuffle death star dodgy dealings shady agreements never far away would say main appeal show due fact go show would dare forget pretty picture paint mainstream audiences forget charm forget romance oz mess around first episode ever saw strike nasty surreal could say ready watch develop taste oz get accustom high level graphic violence violence injustice crook guard sell nickel inmates kill order get away well mannered middle class inmates turn prison bitch due lack street skills prison experience watch oz may

<h2>Problem- 2</h2>

<h5>Total number of words</h5>

In [36]:
total_words=sum(len(review) for review in df['review'])
print("Total words in corpus:", total_words)

Total words in corpus: 39297350


<h5>Unique words (Vocabulary)</h5>

In [37]:
vocabulary=set()

for review in df['review']:
    vocabulary.update(review)

print("Vocabulary Size:",len(vocabulary))

Vocabulary Size: 127


In [42]:
#saving the df, so i can continue from here later on.
df.to_pickle("imdb_preprocessed.pkl")

In [2]:
df=pd.read_pickle("imdb_preprocessed.pkl")

<h2>Problem- 3</h2>

<h5>Applying One Hot Encoding (OHE)</h5>

In [4]:
ohe=pd.get_dummies(df['sentiment'],dtype=int)
df=pd.concat([df,ohe],axis=1)

In [5]:
df.sample(5)

,review,sentiment,negative,positive
8542,eye labyrinth average giallo honest really sur...,positive,0,1
6805,think unfortunate lead comment movie include w...,positive,0,1
5882,scold scorn fellow christians state disappoint...,negative,1,0
48648,first f117 high tech fighter aircraft secondly...,negative,1,0
6845,quite possibly one greatest waste celluloid pa...,negative,1,0


<h2>Problem- 4</h2>

<h5>Applying Bag of Words</h5>

In [6]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer()

In [7]:
bow=cv.fit_transform(df['review'])

In [9]:
print(cv.vocabulary_)

{'one': 57917, 'reviewers': 67475, 'mention': 51733, 'watch': 87657, 'oz': 59080, 'episode': 26592, 'hook': 38060, 'right': 67811, 'exactly': 27376, 'happen': 35832, 'first': 29619, 'thing': 80522, 'strike': 76984, 'brutality': 11925, 'unflinching': 84340, 'scenes': 70374, 'violence': 86688, 'set': 71766, 'word': 89368, 'go': 33373, 'trust': 82803, 'show': 72754, 'faint': 28143, 'hearted': 36476, 'timid': 81066, 'pull': 64395, 'punch': 64434, 'regard': 66446, 'drug': 24280, 'sex': 71851, 'hardcore': 35884, 'classic': 16003, 'use': 85455, 'call': 12904, 'nickname': 56071, 'give': 33109, 'oswald': 58527, 'maximum': 50734, 'security': 71197, 'state': 76169, 'penitentary': 60490, 'focus': 30184, 'mainly': 49306, 'emerald': 25935, 'city': 15879, 'experimental': 27674, 'section': 71181, 'prison': 63564, 'cells': 14233, 'glass': 33176, 'front': 31223, 'face': 28033, 'inwards': 41321, 'privacy': 63580, 'high': 37236, 'agenda': 2859, 'home': 37870, 'many': 49902, 'aryans': 5871, 'muslims': 5472

<h2>Problem- 5</h2>

<h5>bi-gram</h5
               >

In [12]:
n_grams=CountVectorizer(ngram_range=(2,2))

In [13]:
ng=n_grams.fit_transform(df['review'])

In [17]:
print("Bigram vocabulary size:", len(n_grams.vocabulary_))

Bigram vocabulary size: 2740881


<h5>tri-gram</h5>

In [15]:
n_grams2=CountVectorizer(ngram_range=(3,3))
ng2=n_grams2.fit_transform(df['review'])

In [18]:
print("Trigram vocabulary size:", len(n_grams2.vocabulary_))

Trigram vocabulary size: 5311757


<p>Observation: The vocabulary dimensionality increases significantly as the value of N increases. The bigram vocabulary contains 2,740,881 unique combinations, while the trigram vocabulary contains 5,311,757 unique combinations. Thus, higher-order n-grams result in higher dimensionality because they generate more unique word combinations.</p>

<h2>Problem- 6</h2>

<h5>Tf-Idf</h5>

In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidfv=TfidfVectorizer()

In [21]:
tfidf=tfidfv.fit_transform(df['review'])

In [22]:
print("TfIdf vocabulary size:", len(tfidfv.vocabulary_))

TfIdf vocabulary size: 91071


In [23]:
#IDF SCORE DATAFRAME
idf_df=pd.DataFrame({
    'word':tfidfv.get_feature_names_out(),
    'idf_val':tfidfv.idf_
})

In [24]:
idf_df.head()

,word,idf_val
0,00,6.638015
1,000,5.703906
2,00000000000,11.126651
3,0000000000001,11.126651
4,00000001,11.126651
